gengo/src/lib.rs

```diff
@@ -39,127 +42,243 @@ const GLOB_MATCH_OPTIONS: MatchOptions = MatchOptions {
+                            let mut abs_root = root.clone();
+                            if !abs_root.is_empty() {
+                                abs_root.push(b'/');
+                            }
+                            abs_root.extend_from_slice(&path);
+                            Some((abs_root, sm_repo, tree_id))
+                        }),
+                );
+            }
+        }
+
+        Ok(Analysis(all_results))
     }
 
-    fn analyze_tree(
+    fn analyze_index(
         &self,
-        root: &str,
-        tree: &Tree,
-        results: &mut IndexMap<PathBuf, Entry>,
+        repo: &gix::ThreadSafeRepository,
+        results: &mut Results,
+        state: GitState,
     ) -> Result<()> {
-        for entry in tree.iter() {
-            let kind = entry.kind();
-            // HACK Skip submodules. Might want to refactor this later.
-            let object = if let Some(ObjectType::Commit) = kind {
-                // NOTE This is a submodule.
-                continue;
-            } else {
-                entry.to_object(&self.repository)?
-            };
-            match kind {
-                Some(ObjectType::Tree) => {
-                    let path = entry.name().ok_or("invalid path")?;
-                    let tree = object.as_tree().expect("object to be a tree");
-                    let path = Path::new(root).join(path);
-                    let path = path.to_str().ok_or("invalid path")?;
-
-                    self.analyze_tree(path, tree, results)?;
-                }
-                Some(ObjectType::Blob) => {
-                    let path = entry.name().ok_or("invalid path").unwrap();
-                    let filepath = Path::new(root).join(path);
-                    let filepath = filepath.as_os_str();
-                    let blob = object.as_blob().expect("object to be a blob");
-
-                    self.analyze_blob(filepath, blob, results)?;
+        gix::parallel::in_parallel_with_slice(
+            &mut results.entries,
+            None,
+            move |_| (state.clone(), repo.to_thread_local()),
+            |entry, (state, repo), _, should_interrupt| {
+                if should_interrupt.load(Ordering::Relaxed) {
+                    return Ok(());
                 }
-                _ => continue,
-            }
-        }
+                let Ok(path) =
+                    gix::path::try_from_bstr(entry.index_entry.path_in(&results.path_storage))
+                else {
+                    return Ok(());
+                };
+                self.analyze_blob(path, repo, state, entry)
+            },
+            || Some(std::time::Duration::from_micros(5)),
+            std::convert::identity,
+        )?;
         Ok(())
     }
 
-    fn analyze_blob<P: AsRef<Path>>(
+    fn analyze_blob(
         &self,
-        filepath: P,
-        blob: &Blob,
-        results: &mut IndexMap<PathBuf, Entry>,
+        filepath: impl AsRef<Path>,
```

Comment:
Just so I can learn: why switch from a generic type to `impl AsRef<Path>`?

Reply:
It's less noisy to my eyes and a syntax that was added to provide a reasonable alternative to the one it (partially) replaces. I'd consider it more idiomatic, too.